File: reverse-linkedlist-recursive.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


#Gives wrong answer on leetcode - my outptut is always empty list []
#Time complexity - O(n)
#Space complexity - O(1)

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None:
            return None
        def reversehelper(curr, prev):
            if curr.next == None:
                curr.next = prev
                head = curr
                return head
            next = curr.next
            curr.next = prev
            reversehelper(next, curr)
        return reversehelper(head, None)

File: ReverseLinkedList_Recursive.java
// Time Complexity : O(n)
// Space Complexity : O(n) 
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        //null check
        if(head == null || head.next == null) return head;
        
        ListNode reversed = reverseList(head.next);
        //st.pop()
        //head is coming out of stack
        
        head.next.next = head;
        head.next = null;
        return reversed;
        
    }
}